src/coreclr/jit/gentree.cpp

```diff
@@ -2620,29 +2620,58 @@ unsigned Compiler::gtSetMultiOpOrder(GenTreeMultiOp* multiOp)
     unsigned level  = 0;
     unsigned lvl2   = 0;
 
-#if defined(FEATURE_HW_INTRINSICS) && defined(TARGET_XARCH)
-    if (multiOp->OperIs(GT_HWINTRINSIC) && (multiOp->GetOperandCount() == 1) &&
-        multiOp->AsHWIntrinsic()->OperIsMemoryLoadOrStore())
+#if defined(FEATURE_HW_INTRINSICS)
+    if (multiOp->OperIs(GT_HWINTRINSIC))
     {
-        costEx = IND_COST_EX;
-        costSz = 2;
+        GenTreeHWIntrinsic* hwTree = multiOp->AsHWIntrinsic();
+#if defined(TARGET_XARCH)
+        if ((hwTree->GetOperandCount() == 1) && hwTree->OperIsMemoryLoadOrStore())
+        {
+            costEx = IND_COST_EX;
+            costSz = 2;
 
-        GenTree* addr = multiOp->Op(1)->gtEffectiveVal();
-        level         = gtSetEvalOrder(addr);
+            GenTree* addr = hwTree->Op(1)->gtEffectiveVal();
+            level         = gtSetEvalOrder(addr);
 
-        // See if we can form a complex addressing mode.
-        if (addr->OperIs(GT_ADD) && gtMarkAddrMode(addr, &costEx, &costSz, multiOp->TypeGet()))
-        {
-            // Nothing to do, costs have been set.
+            // See if we can form a complex addressing mode.
+            if (addr->OperIs(GT_ADD) && gtMarkAddrMode(addr, &costEx, &costSz, hwTree->TypeGet()))
+            {
+                // Nothing to do, costs have been set.
+            }
+            else
+            {
+                costEx += addr->GetCostEx();
+                costSz += addr->GetCostSz();
+            }
+
+            hwTree->SetCosts(costEx, costSz);
+            return level;
         }
-        else
+#endif
+        switch (hwTree->GetHWIntrinsicId())
         {
-            costEx += addr->GetCostEx();
-            costSz += addr->GetCostSz();
+#if defined(TARGET_XARCH)
+            case NI_Vector128_Create:
+            case NI_Vector256_Create:
+#elif defined(TARGET_ARM64)
+            case NI_Vector64_Create:
+            case NI_Vector128_Create:
+#endif
+            {
+                if ((hwTree->GetOperandCount() == 1) && hwTree->Op(1)->OperIsConst())
```

Comment:
Why are we only doing `OperandCount == 1`?

What about the cases where `OperandCount == 2` through `OperandCount == 32`? Are those being properly tracked as "expensive" and getting CSEd?

Reply:
@tannergooding yeah, they are assigned a higher cost automatically due to multiple arguments so the problem doesn't reproduce for them. but that's a good point, I guess Vector128.Create(1,2,3,4,5,6,7,8) currently gets a very high cost while in reality it should still be 3/2

Reply:
I think its fine to log an issue for and cover in a separate issue here.